rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,571 @@
+- `glidesId` (string): unique value representing this trip for the given `serviceDate`
+```json
+{"serviceDate": "2023-01-20", "glidesId": "ADDED-123"}
+```
+**Scheduled Trip**
+- `serviceDate` (RFC3999 date): the service date for the trip
+- `startLocation` (Location): where the trip is scheduled to start
+- `endLocation` (Location): where the trip is scheduled to end
+- `departureTime` (Time): when the trip is scheduled to depart `startLocation`
+- `arrivalTime` (Time): when the trip is scheduled to arrive at `endLocation`
+
+```json
+{
+  "serviceDate": "2023-01-20",
+  "startLocation": {"gtfsId": "place-mdftf"},
+  "endLocation": {"gtfsId": "place-heath"},
+  "departureTime": "4:47:00",
+  "arrivalTime": "5:34:00"
+}
+```
+*Notes*:
+- The added/scheduled distinction here is internal to Glides, and may not align with what is in other data sources. For example, an Added Trip in Glides may still correspond to a scheduled GTFS trip, and a Scheduled Trip may not appear in GTFS (due to track closures, for example)
+
+## Events
+### Batched
+If multiple events are generated as a part of the same author action, then they can be combined into a single Batched event. This ensures that all of the events are kept together throughout the data pipeline.
+
+Event type: `com.mbta.ctd.glides.batched`
+Fields in the event:
+- `author` (Author, optional): the inspector who performed the events, if it is the same for all the events in the `events` array. 
+- `events` (array of `{"type": string, "data": JSON}`): each element in the array is a CloudEvent including only the `type` and `data` field, and the `data` fields of the individual events do not need to re-include the `author` field if it is the same as the top-level `author`. See [Examples][#examples] for an example.
+
+### Imported date schedules
+These values map a service date to a schedule identifier for the given line. This event may happen at any time, and subsequent events will update the `scheduleId` used on the `serviceDate` for the `line` (but not the `scheduleId` used on the `serviceDate` for other lines).
+
+Event type: `com.mbta.ctd.glides.imported-date-schedules`
+Fields in the event:
+- `line` (string): the line for which a new date/schedule mapping is being imported
+- `ratingName` (string): the rating to which the new dates belong
+- `schedules`: Array of:
+	- `serviceDate` (RFC3999 date): date on which the given schedule will be used
+	- `scheduleId` (string): unique identifier for the schedule of trips running on this date 
+
+### Imported runs
+Runs map an operator to their work for a given `serviceDate`. Subsequent events update the `operator` assigned to the `runId` on the given `serviceDate`.
+
+Event type: `com.mbta.ctd.glides.imported-runs`
+Fields in the event:
+- `runs`: Array of:
+	- `serviceDate` (RFC3999 date)
+	- `runId` (string): unique identifier for the run. Must be unique for the given `serviceDate`.
+	- `operator` (Operator)
+
+### Imported scheduled trips
```

Comment:
It'd require an extra "click this button to broadcast the newly imported rating on the Kinesis stream" step in the import process, which isn't ideal but could be done.

One downside of any approach that broadcasts the schedule at import time is that ratings last roughly three months. Broadcasting and then consuming three months worth of data at once might be somewhat taxing, especially given the low throughput that the stream will have at every other time, and looking back three months to see the scheduled operator for a trip on the last day of the rating might be architecturally complicated.

Adding `scheduledOperator` to `trip_updated` avoids that downside.

Another solution would be to broadcast a single event for just one day at a time overnight before that day; this seems like it's close to how heavy rail trainsheets get loaded into RTR. (It would mean that when an inspector adjusts the next day's runs, consumers don't yet have the scheduled operators for those trips.)

Reply:
I've added these options (and this issue in general) to the "Unresolved questions" for us to discuss later.

Reply:
84a9be343fd03649f330eee0f387d22b431f2681